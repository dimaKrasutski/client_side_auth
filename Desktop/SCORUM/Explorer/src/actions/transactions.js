import { createActions } from 'redux-actions';

export const {
  setTransactions,
  setTransactionsOffset,
  selectTransactionsTab,
  setOperations,
  setOperationsOffset,
} = createActions(
  'SET_TRANSACTIONS',
  'SET_TRANSACTIONS_OFFSET',
  'SELECT_TRANSACTIONS_TAB',
  'SET_OPERATIONS',
  'SET_OPERATIONS_OFFSET',
);
