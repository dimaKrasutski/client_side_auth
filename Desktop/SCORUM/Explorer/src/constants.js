import * as THEME_CONSTANTS from 'theme/constants';

export const {
  LOCALE_KEYS,
  LANG_KEYS,
  LOCALE,
  LOCALES,
  DEFAULT_LOCALE,
  DEFAULT_LANG,
} = THEME_CONSTANTS;

export const UNDEFINED_ERROR_ID = 'explorer.error.unknown';

export const MIN_FRACTION_DIGITS = 3;
export const MAX_FRACTION_DIGITS = 9;

export const CURRENCY_USD = 'USD';
export const CURRENCY_EUR = 'EUR';
export const ALL_CURRENCY = [CURRENCY_USD, CURRENCY_EUR];
