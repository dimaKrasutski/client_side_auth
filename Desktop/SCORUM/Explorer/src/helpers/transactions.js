import { compose, nth, over, lensProp, pick, prop, last, dec, map, reverse } from 'ramda';
import { Subject } from 'rxjs/Subject';
import { getAccountTransactions } from '../api/scorum';

const INITIAL_HISTORY_OFFSET = -1;

export const TRANSACTION_OP_FIELD_NAME = 'op';
export const TRANSACTION_TIMESTAMP_FIELD_NAME = 'timestamp';
export const TRANSACTION_AMOUNT_FIELD_NAME = 'amount';
export const TRANSACTION_FROM_FIELD_NAME = 'from';
export const TRANSACTION_TO_FIELD_NAME = 'to';
export const TRANSACTION_MEMO_FIELD_NAME = 'memo';
export const TRANSACTION_TRX_ID_FIELD_NAME = 'trx_id';
export const TRANSACTION_BLOCK_FIELD_NAME = 'block';

/**
 * Get first transaction from a list and take index property decrementing it
 */
export const getNextOffset = compose(dec, prop('index'), last);

/**
 * @param {[]} transaction - user history transaction
 * @returns {{}} - reduced transaction
 */
export const reduceHistoryTransaction = compose(
  // op field constains transaction type (we don't need) and data object itself.
  over(lensProp(TRANSACTION_OP_FIELD_NAME), nth(1)),
  transaction => ({
    index: transaction[0],
    ...pick([ // spread whitelisted keys object
      TRANSACTION_TIMESTAMP_FIELD_NAME,
      TRANSACTION_BLOCK_FIELD_NAME,
      TRANSACTION_TRX_ID_FIELD_NAME,
      TRANSACTION_OP_FIELD_NAME,
    ], transaction[1]),
  }),
);

/**
 * @param {{}} conversionTransaction - reduced transaction
 * @returns {{}} - reduced conversion
 */
export const reduceConversion = compose(
  // Mapping op property leaving only amount
  over(lensProp(TRANSACTION_OP_FIELD_NAME), pick([TRANSACTION_AMOUNT_FIELD_NAME])),
  reduceHistoryTransaction,
);

/**
 * @param {{}} conversionTransaction - reduced transaction
 * @returns {{}} - reduced conversion
 */
export const reduceTransaction = compose(
  // Mapping op property leaving only amount
  over(
    lensProp(TRANSACTION_OP_FIELD_NAME),
    pick([
      TRANSACTION_AMOUNT_FIELD_NAME,
      TRANSACTION_FROM_FIELD_NAME,
      TRANSACTION_TO_FIELD_NAME,
      TRANSACTION_MEMO_FIELD_NAME,
    ]),
  ),
  reduceHistoryTransaction,
);

/**
 * @param {function} reducer - function to reduce entity items
 * @param {string} username - name of user to fetch history items for
 * @param {number} initialLimit - fetch conversions by portions, sliced with limit
 */
export const createTransactionsFetcher = (reducer, username, initialLimit) => {
  const transactionsSubject = new Subject();
  let offset = INITIAL_HISTORY_OFFSET;
  let limit = initialLimit;
  let isStopped = false;
  let inProgress = false;

  transactionsSubject.isInitial = () => offset === INITIAL_HISTORY_OFFSET;

  transactionsSubject.hasNext = () => !isStopped;

  transactionsSubject.fetchNext = function fetchNext() {
    if (isStopped || inProgress) {
      return null;
    }

    inProgress = true;
    return getAccountTransactions(username, offset, limit)
      .subscribe(compose(
        (fetched) => {
          this.next(fetched);

          offset = getNextOffset(fetched);
          if (limit > offset && offset >= 1) { // Blockchain doesn't allow limit bigger than offset
            limit = offset;
          }
          if (Number.isNaN(offset) || offset < 1) { // When we trap to negative - no more records left
            isStopped = true;
          }
          inProgress = false;
        },
        reverse,
        map(reducer)
      ));
  };

  transactionsSubject.resetAndFetchNext = function resetAndFetchNext() {
    offset = INITIAL_HISTORY_OFFSET;
    limit = initialLimit;
    isStopped = false;
    inProgress = false;
    this.fetchNext();
  };

  return transactionsSubject;
};
