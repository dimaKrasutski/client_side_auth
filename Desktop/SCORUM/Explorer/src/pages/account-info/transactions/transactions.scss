.transactions-table {
  .fixed-column {
    width: 28%;
  }
}
