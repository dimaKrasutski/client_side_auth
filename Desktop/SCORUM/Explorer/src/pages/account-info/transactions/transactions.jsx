import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { FormattedMessage } from 'react-intl';
import Table from 'reactstrap/lib/Table';
import { Transaction } from './transaction/transaction';
import { PreloaderTransactions } from './preloader/preloader-transactions';

import './transactions.scss';

export const TransactionsUI = ({
  transactions, inProgress, locale, userName,
}) => (
  <Table className="mt-3 transactions-table">
    <thead className="thead-light">
      <tr>
        <th className="fixed-column">
          <FormattedMessage id="explorer.account-info.transactions.hash" />
        </th>
        <th>
          <FormattedMessage id="explorer.account-info.transactions.who" />
        </th>
        <th>
          <FormattedMessage id="explorer.account-info.transactions.date" />
        </th>
        <th className="text-right">
          <FormattedMessage id="explorer.account-info.transactions.amount" />
        </th>
      </tr>
    </thead>
    <tbody>
      {transactions === null ?
        <PreloaderTransactions /> :
        transactions.map(transaction => (
          <Transaction key={transaction.index} locale={locale} transaction={transaction} userName={userName} />
        ))
      }
      {inProgress && <PreloaderTransactions />}
    </tbody>
  </Table>
);

TransactionsUI.propTypes = {
  userName: PropTypes.string.isRequired,
  locale: PropTypes.string.isRequired,
  transactions: PropTypes.arrayOf(PropTypes.shape()),
  inProgress: PropTypes.bool.isRequired,
};

TransactionsUI.defaultProps = {
  transactions: null,
};

export const mapStateToProps = ({ intl }) => ({
  locale: intl.localeFull,
});

export const Transactions = connect(mapStateToProps)(TransactionsUI);
