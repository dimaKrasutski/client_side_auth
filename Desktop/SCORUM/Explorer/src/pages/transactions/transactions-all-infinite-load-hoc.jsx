import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'ramda';
import { createTransactionsAllFetcher } from '../../helpers/operations';
import { withInfiniteScroll } from '../../hocs/with-infinite-scroll';
import { setOperations, setOperationsOffset } from '../../actions/transactions';

export const withTransactionsAllInfiniteLoad = (WrappedComponent) => {
  const WrappedWithInfiniteScroll = withInfiniteScroll(WrappedComponent);

  class TransactionsAllInfiniteLoad extends Component {
    static propTypes = {
      operationsOffset: PropTypes.number.isRequired,
      isValidTabSelected: PropTypes.bool.isRequired,
      operations: PropTypes.arrayOf().isRequired,
      setOperationsAction: PropTypes.func.isRequired,
      setOperationsOffsetAction: PropTypes.func.isRequired,
    };

    transactionsFetcher = createTransactionsAllFetcher(
      this.props.operationsOffset,
      this.props.setOperationsOffsetAction,
    );

    transactionsFetcherSubscription;

    state = {
      inProgress: false,
    };

    /**
     * Initial subscribe to fetcher and instant preload of first bunch
     */
    componentDidMount() {
      const { operationsOffset } = this.props;
      this.transactionsFetcherSubscription = this.transactionsFetcher.subscribe(this.updateTransactions);
      if (operationsOffset === -1) {
        this.transactionsFetcher.fetchNext();
      }
    }

    componentWillUnmount() {
      this.transactionsFetcherSubscription.unsubscribe();
    }

    /**
     * Initiate loading of next transactions
     */
    loadNext = () => {
      if (this.props.isValidTabSelected && this.transactionsFetcher.hasNext()) {
        this.setState({
          inProgress: true,
        });
        this.transactionsFetcher.fetchNext();
      }
    };

    updateTransactions = (nextTransactions) => {
      this.props.setOperationsAction(nextTransactions);
      this.setState({
        inProgress: false,
      });
    };

    render() {
      const { inProgress } = this.state;

      return (
        <WrappedWithInfiniteScroll
          {...this.props}
          inProgress={inProgress}
          loadNext={this.loadNext}
        />
      );
    }
  }

  const mapStateToProps = ({ transactions }) => ({
    operations: transactions.operations,
    isValidTabSelected: transactions.selectedTab === 'all',
    operationsOffset: transactions.operationsOffset,
  });

  const mapDispatchToProps = dispatch => ({
    setOperationsAction: compose(dispatch, setOperations),
    setOperationsOffsetAction: compose(dispatch, setOperationsOffset),
  });

  return connect(mapStateToProps, mapDispatchToProps)(TransactionsAllInfiniteLoad);
};
