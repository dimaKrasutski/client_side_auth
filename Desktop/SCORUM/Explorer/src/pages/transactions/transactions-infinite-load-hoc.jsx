import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'ramda';
import { createTransactionsFetcher } from '../../helpers/operations';
import { withInfiniteScroll } from '../../hocs/with-infinite-scroll';
import { setTransactions, setTransactionsOffset } from '../../actions/transactions';

export const withTransactionsInfiniteLoad = (WrappedComponent) => {
  const WrappedWithInfiniteScroll = withInfiniteScroll(WrappedComponent);

  class TransactionsInfiniteLoad extends Component {
    static propTypes = {
      transactionsOffset: PropTypes.number.isRequired,
      transactions: PropTypes.arrayOf().isRequired,
      isValidTabSelected: PropTypes.bool.isRequired,
      setTransactionsAction: PropTypes.func.isRequired,
      setTransactionsOffsetActions: PropTypes.func.isRequired,
    };

    transactionsFetcher = createTransactionsFetcher(
      this.props.transactionsOffset,
      this.props.setTransactionsOffsetActions
    );

    transactionsFetcherSubscription;

    state = {
      inProgress: false,
    };

    /**
     * Initial subscribe to fetcher and instant preload of first bunch
     */
    componentDidMount() {
      const { transactionsOffset } = this.props;
      this.transactionsFetcherSubscription = this.transactionsFetcher.subscribe(this.updateTransactions);
      if (transactionsOffset === -1) {
        this.transactionsFetcher.fetchNext();
      }
    }

    componentWillUnmount() {
      this.transactionsFetcherSubscription.unsubscribe();
    }

    /**
     * Initiate loading of next transactions
     */
    loadNext = () => {
      if (this.props.isValidTabSelected && this.transactionsFetcher.hasNext()) {
        this.setState({
          inProgress: true,
        });
        this.transactionsFetcher.fetchNext();
      }
    };

    updateTransactions = (nextTransactions) => {
      this.props.setTransactionsAction(nextTransactions);
      this.setState({
        inProgress: false,
      });
    };

    render() {
      const { inProgress } = this.state;

      return (
        <WrappedWithInfiniteScroll
          {...this.props}
          inProgress={inProgress}
          loadNext={this.loadNext}
        />
      );
    }
  }

  const mapStateToProps = ({ transactions }) => ({
    transactions: transactions.transactions,
    isValidTabSelected: transactions.selectedTab === 'transactions',
    transactionsOffset: transactions.transactionsOffset,
  });

  const mapDispatchToProps = dispatch => ({
    setTransactionsAction: compose(dispatch, setTransactions),
    setTransactionsOffsetActions: compose(dispatch, setTransactionsOffset),
  });

  return connect(mapStateToProps, mapDispatchToProps)(TransactionsInfiniteLoad);
};
