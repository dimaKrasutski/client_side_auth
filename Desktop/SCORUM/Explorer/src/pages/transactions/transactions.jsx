import React from 'react';
import { connect } from 'react-redux';
import { Container, Row, Col } from 'reactstrap';
import { Tabs, Tab } from 'theme/components/tabs/tabs';
import PropTypes from 'prop-types';
import { compose } from 'ramda';
import { PageTitle } from '../../components/page-title/page-title';
import { Main } from '../main';
import { TransactionsTable } from './transactions-table/transactions-table';
import { OperationsTable } from './operations-table/operations-table';
import { withTransactionsInfiniteLoad } from './transactions-infinite-load-hoc';
import { withTransactionsAllInfiniteLoad } from './transactions-all-infinite-load-hoc';
import { selectTransactionsTab } from '../../actions/transactions';

const TransactionsWithInfiniteLoad = withTransactionsInfiniteLoad(TransactionsTable);
const TransactionsAllWithInfiniteLoad = withTransactionsAllInfiniteLoad(OperationsTable);

export const TransactionsUI = ({
  selectTab,
  selectedTab,
}) => (
  <Main>
    <Container className="pt-6 operations-container">
      <Row>
        <Col xs="12">
          <PageTitle title="explorer.transactions.title" />
        </Col>
      </Row>
      <Row>
        <Col xs="12">
          <Tabs onTabSelect={selectTab} activeTab={selectedTab}>
            <Tab tabId="transactions" label="explorer.block-info.transactions.title">
              <TransactionsWithInfiniteLoad />
            </Tab>
            <Tab tabId="all" label="explorer.block-info.transactions.all">
              <TransactionsAllWithInfiniteLoad />
            </Tab>
          </Tabs>
        </Col>
      </Row>
    </Container>
  </Main>
);

TransactionsUI.propTypes = {
  selectedTab: PropTypes.string.isRequired,
  selectTab: PropTypes.func.isRequired,
};

const mapStateToProps = ({ transactions }) => ({
  selectedTab: transactions.selectedTab,
});

const mapDispatchToProps = dispatch => ({
  selectTab: compose(dispatch, selectTransactionsTab),
});

export const Transactions = connect(mapStateToProps, mapDispatchToProps)(TransactionsUI);

