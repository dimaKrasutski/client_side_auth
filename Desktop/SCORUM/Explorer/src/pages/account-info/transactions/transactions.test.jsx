import React from 'react';
import { TransactionsUI, mapStateToProps } from './transactions';
import { DEFAULT_LOCALE } from '../../../constants';

const transactions = [
  {
    timestamp: '2018-03-07T15:42:39',
    index: 0,
    op: {
      amount: '100000.000 SCR',
    },
  },
  {
    timestamp: '2018-03-07T15:42:41',
    index: 1,
    op: {
      amount: '100000.000 SCR',
    },
  },
  {
    timestamp: '2018-03-07T15:42:49',
    index: 2,
    op: {
      amount: '100000.000 SCR',
    },
  },
];

const loadNextMock = jest.fn();
let wrapper;

describe('<TransactionsUI />', () => {
  beforeEach(() => {
    loadNextMock.mockReset();
    wrapper = shallow(<TransactionsUI locale={DEFAULT_LOCALE} userName="Dabooos" inProgress={false} transactions={transactions} loadNext={loadNextMock} />);
  });

  test('should', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should match with inProgress', () => {
    wrapper = shallow(<TransactionsUI locale={DEFAULT_LOCALE} userName="Dabooos" transactions={transactions} inProgress loadNext={loadNextMock} />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('mapStateToProps', () => {
  test('should fill props from state', () => {
    const state = {
      intl: {
        localeFull: 'ko-ko',
      },
    };
    expect(mapStateToProps(state).locale).toEqual('ko-ko');
  });
});
