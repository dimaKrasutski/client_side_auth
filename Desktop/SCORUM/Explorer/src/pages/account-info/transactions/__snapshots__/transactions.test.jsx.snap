// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TransactionsUI /> should 1`] = `
<Table
  className="mt-3 transactions-table"
  responsiveTag="div"
  tag="table"
>
  <thead
    className="thead-light"
  >
    <tr>
      <th
        className="fixed-column"
      >
        <FormattedMessage
          id="explorer.account-info.transactions.hash"
          values={Object {}}
        />
      </th>
      <th>
        <FormattedMessage
          id="explorer.account-info.transactions.who"
          values={Object {}}
        />
      </th>
      <th>
        <FormattedMessage
          id="explorer.account-info.transactions.date"
          values={Object {}}
        />
      </th>
      <th
        className="text-right"
      >
        <FormattedMessage
          id="explorer.account-info.transactions.amount"
          values={Object {}}
        />
      </th>
    </tr>
  </thead>
  <tbody>
    <Connect(TransactionUI)
      key="0"
      locale="en-us"
      transaction={
        Object {
          "index": 0,
          "op": Object {
            "amount": "100000.000 SCR",
          },
          "timestamp": "2018-03-07T15:42:39",
        }
      }
      userName="Dabooos"
    />
    <Connect(TransactionUI)
      key="1"
      locale="en-us"
      transaction={
        Object {
          "index": 1,
          "op": Object {
            "amount": "100000.000 SCR",
          },
          "timestamp": "2018-03-07T15:42:41",
        }
      }
      userName="Dabooos"
    />
    <Connect(TransactionUI)
      key="2"
      locale="en-us"
      transaction={
        Object {
          "index": 2,
          "op": Object {
            "amount": "100000.000 SCR",
          },
          "timestamp": "2018-03-07T15:42:49",
        }
      }
      userName="Dabooos"
    />
  </tbody>
</Table>
`;

exports[`<TransactionsUI /> should match with inProgress 1`] = `
<Table
  className="mt-3 transactions-table"
  responsiveTag="div"
  tag="table"
>
  <thead
    className="thead-light"
  >
    <tr>
      <th
        className="fixed-column"
      >
        <FormattedMessage
          id="explorer.account-info.transactions.hash"
          values={Object {}}
        />
      </th>
      <th>
        <FormattedMessage
          id="explorer.account-info.transactions.who"
          values={Object {}}
        />
      </th>
      <th>
        <FormattedMessage
          id="explorer.account-info.transactions.date"
          values={Object {}}
        />
      </th>
      <th
        className="text-right"
      >
        <FormattedMessage
          id="explorer.account-info.transactions.amount"
          values={Object {}}
        />
      </th>
    </tr>
  </thead>
  <tbody>
    <Connect(TransactionUI)
      key="0"
      locale="en-us"
      transaction={
        Object {
          "index": 0,
          "op": Object {
            "amount": "100000.000 SCR",
          },
          "timestamp": "2018-03-07T15:42:39",
        }
      }
      userName="Dabooos"
    />
    <Connect(TransactionUI)
      key="1"
      locale="en-us"
      transaction={
        Object {
          "index": 1,
          "op": Object {
            "amount": "100000.000 SCR",
          },
          "timestamp": "2018-03-07T15:42:41",
        }
      }
      userName="Dabooos"
    />
    <Connect(TransactionUI)
      key="2"
      locale="en-us"
      transaction={
        Object {
          "index": 2,
          "op": Object {
            "amount": "100000.000 SCR",
          },
          "timestamp": "2018-03-07T15:42:49",
        }
      }
      userName="Dabooos"
    />
    <PreloaderTransactions />
  </tbody>
</Table>
`;
